server/game/gameSystems/ScrySystem.ts

```diff
@@ -0,0 +1,84 @@
+import type { AbilityContext } from '../core/ability/AbilityContext';
+import type { Card } from '../core/card/Card';
+import { CardType, Location, WildcardCardType } from '../core/Constants';
+import { CardTargetSystem, type ICardTargetSystemProperties } from '../core/gameSystem/CardTargetSystem';
+import { GameEvent } from '../core/event/GameEvent';
+import { moveCard } from './GameSystemLibrary';
+
+export interface IScryProperties extends ICardTargetSystemProperties {
+    amount: number;
+}
+
+export class ScrySystem<TContext extends AbilityContext = AbilityContext> extends CardTargetSystem<TContext, IScryProperties> {
+    public override readonly name = 'scry';
+    public override targetTypeFilter = [WildcardCardType.Unit, WildcardCardType.Upgrade, CardType.Event];
+
+    protected override defaultProperties: IScryProperties = {
+        amount: 1,
+    };
+
+    // eslint-disable-next-line @typescript-eslint/no-empty-function
+    public override eventHandler(event): void { }
+
+    public override queueGenerateEventGameSteps(events: GameEvent[], context: TContext): void {
+        const player = context.player;
+        const { amount } = this.generatePropertiesFromContext(context) as IScryProperties;
+        const deckLength = player.drawDeck.length;
+        const actual_amount = amount === -1 ? deckLength : (amount > deckLength ? deckLength : amount);
+        const cards = player.drawDeck.slice(0, actual_amount);
+        let decision_mapper = cards.map((card: Card) => [
+            card.internalName + '-top',
+            card.internalName + '-bottom',
+        ]).reduce((acc, val) => acc.concat(val), []);
+
+        // @ts-ignore
+        const cardHandler = (player, card) => {
+            decision_mapper = decision_mapper.filter((a) => a !== card + '-top' && a !== card + '-bottom');
+            if (decision_mapper.length > 0) {
+                context.game.promptWithHandlerMenu(player, {
+                    activePromptTitle: 'Select card to move to the top or bottom of the deck',
+                    context: context,
+                    choices: decision_mapper,
+                    choiceHandler: (choice) => {
```

Comment:
There's another option here worth considering, which is using `handlers` instead of `choiceHandler`. The way handlers works is that you provide an array of choices (which you already have) and an array of handlers that is the same length. Then if choice index `i` is chosen, handler index `i` is called.

It might take a couple tries to get the indexing right but it would decouple the button names from the implementation, which would be really nice if we ever need to change them

Reply:
transformed from choiceHandler to handlers